scylla/src/transport/session.rs

```diff
@@ -283,8 +283,40 @@ impl Session {
 
         node_addresses.extend(resolved);
 
+        let use_ssl = match () {
+            #[cfg(not(feature = "ssl"))]
+            () => false,
+            #[cfg(feature = "ssl")]
+            () => config.ssl_context.is_some(),
+        };
+
+        let mut shard_aware_addresses: Vec<SocketAddr> = vec![];
+        if let Some(shard_aware_port) =
+            Self::get_shard_aware_port(node_addresses[0], config.get_connection_config(), use_ssl)
+                .await
+        {
+            info!("Shard-aware port detected: {}", shard_aware_port);
+            shard_aware_addresses = (&node_addresses)
+                .iter()
+                .map(|addr| SocketAddr::new(addr.ip(), shard_aware_port))
+                .collect();
+        }
+
         // Start the session
-        let cluster = Cluster::new(&node_addresses, config.get_connection_config()).await?;
+        let shard_aware_cluster = if !shard_aware_addresses.is_empty() {
+            Cluster::new(&shard_aware_addresses, config.get_connection_config())
+                .await
+                .ok()
+        } else {
+            None
+        };
+        let cluster = match shard_aware_cluster {
+            Some(clust) => clust,
+            None => {
+                warn!("Unable to establish connections at detected shard-aware port, falling back to default ports");
```

Comment:
Here, maybe we could print what error happened while attempting to connect to the shard-aware port?

Reply:
Good idea, done. I collapsed these two pattern matching statements into one